include/oneapi/dpl/experimental/kt/two_pass_scan.h

```diff
@@ -199,15 +224,37 @@ two_pass_scan(sycl::queue q, _InRng&& __in_rng, _OutRng&& __out_rng,
                 if (sub_group_id == 0)
                 {
                     start_idx = (g * num_sub_groups_local);
-                    // TODO: handle if num_sub_groups_local < VL
-                    constexpr std::uint8_t iters = num_sub_groups_local / VL;
+                    std::uint8_t iters = oneapi::dpl::__internal::__dpl_ceiling_div(num_sub_groups_local, VL);
                     sub_group_carry = identity;
-                    _ONEDPL_PRAGMA_UNROLL
-                    for (std::uint8_t i = 0; i < iters; i++)
+                    if (is_full_carry_scanner)
                     {
-                        v = sub_group_partials[i * VL + sub_group_local_id];
-                        std::tie(v, sub_group_carry) = sub_group_scan<VL, true>(sub_group, std::move(v), binary_op, std::move(sub_group_carry));
-                        tmp_storage[start_idx + i * VL + sub_group_local_id] = v;
+                        for (std::uint8_t i = 0; i < iters; i++)
+                        {
+                            v = sub_group_partials[i * VL + sub_group_local_id];
+                            std::tie(v, sub_group_carry) = sub_group_scan<VL, true>(sub_group, std::move(v), binary_op, std::move(sub_group_carry));
+                            tmp_storage[start_idx + i * VL + sub_group_local_id] = v;
+                        }
+                    }
+                    else
+                    {
+                        // In practice iters is usually 1 here when the number of sub-groups is less than the vector length.
+                        // An exception would be the unlikely case where the sub-group size does not divide the work-group size
+                        std::uint8_t i = 0;
+                        for (; i < iters - 1; i++)
+                        {
+                            v = sub_group_partials[sub_group_local_id];
+                            std::tie(v, sub_group_carry) = sub_group_scan<VL, true>(sub_group, std::move(v), binary_op,
+                                                                                    std::move(sub_group_carry));
+                            tmp_storage[start_idx + i * VL + sub_group_local_id] = v;
+                        }
+                        // If we are past the input range, then the previous value of v is passed to the sub-group scan.
+                        // It does not affect the result as our sub_group_scan will use a mask to only process in-range elements.
+                        if (i * VL + sub_group_local_id < num_sub_groups_local)
+                            v = sub_group_partials[sub_group_local_id];
```

Comment:
Same comment here, is there a reason not to add `i * VL` here?

Reply:
It should be here as well.

Reply:
It should be here as well.